// @aurorajs.dev/core/dist/domain/persistence/sql-statement/operator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Operator = void 0;
var Operator;
(function (Operator) {
    Operator["and"] = "[and]";
    Operator["or"] = "[or]";
    Operator["col"] = "[col]";
    Operator["join"] = "[join]";
    Operator["startsWith"] = "[startsWith]";
    Operator["endsWith"] = "[endsWith]";
    Operator["substring"] = "[substring]";
    Operator["not"] = "[not]";
    Operator["is"] = "[is]";
    Operator["in"] = "[in]";
    Operator["notIn"] = "[notIn]";
    Operator["eq"] = "[eq]";
    Operator["ne"] = "[ne]";
    Operator["gte"] = "[gte]";
    Operator["gt"] = "[gt]";
    Operator["lte"] = "[lte]";
    Operator["lt"] = "[lt]";
    Operator["between"] = "[between]";
    Operator["notBetween"] = "[notBetween]";
    Operator["like"] = "[like]";
    Operator["notLike"] = "[notLike]";
    Operator["regexp"] = "[regexp]";
    Operator["notRegexp"] = "[notRegexp]";
    Operator["all"] = "[all]";
    Operator["values"] = "[values]";
    Operator["placeholder"] = "[placeholder]";
    Operator["iLike"] = "[iLike]";
    Operator["notILike"] = "[notILike]";
    Operator["iRegexp"] = "[iRegexp]";
    Operator["notIRegexp"] = "[notIRegexp]";
    Operator["overlap"] = "[overlap]";
    Operator["contains"] = "[contains]";
    Operator["contained"] = "[contained]";
    Operator["adjacent"] = "[adjacent]";
    Operator["strictLeft"] = "[strictLeft]";
    Operator["strictRight"] = "[strictRight]";
    Operator["noExtendRight"] = "[noExtendRight]";
    Operator["noExtendLeft"] = "[noExtendLeft]";
    Operator["any"] = "[any]";
})(Operator = exports.Operator || (exports.Operator = {}));
