// @aurorajs.dev/core/dist/domain/aurora.types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FormatLangCode = exports.AuditingSideEffectEvent = void 0;
var AuditingSideEffectEvent;
(function (AuditingSideEffectEvent) {
    AuditingSideEffectEvent["CREATED"] = "CREATED";
    AuditingSideEffectEvent["BULK_CREATED"] = "BULK_CREATED";
    AuditingSideEffectEvent["UPDATED"] = "UPDATED";
    AuditingSideEffectEvent["BULK_UPDATED"] = "BULK_UPDATED";
    AuditingSideEffectEvent["DELETED"] = "DELETED";
    AuditingSideEffectEvent["BULK_DELETED"] = "BULK_DELETED";
    AuditingSideEffectEvent["RESTORED"] = "RESTORED";
    AuditingSideEffectEvent["BULK_RESTORED"] = "BULK_RESTORED";
    AuditingSideEffectEvent["UPSERTED"] = "UPSERTED";
})(AuditingSideEffectEvent = exports.AuditingSideEffectEvent || (exports.AuditingSideEffectEvent = {}));
var FormatLangCode;
(function (FormatLangCode) {
    FormatLangCode["ID"] = "id";
    FormatLangCode["IETF"] = "ietf";
    FormatLangCode["ISO6392"] = "iso6392";
    FormatLangCode["ISO6393"] = "iso6393";
})(FormatLangCode = exports.FormatLangCode || (exports.FormatLangCode = {}));
